File: src/frontend/src/types/bookmark.ts
/**
 * Bookmark type definitions for the Bookmark Manager application.
 * These interfaces mirror the Go backend model structures.
 */

/**
 * Represents a single bookmark with all its metadata.
 */
export interface Bookmark {
  id: string;
  url: string;
  title: string;
  description?: string;
  tags: string[];
  createdAt: string;
  updatedAt: string;
}

/**
 * Input data for creating a new bookmark.
 * Used in form submissions.
 */
export interface CreateBookmarkInput {
  url: string;
  title: string;
  description?: string;
  tags?: string[];
}

/**
 * Input data for updating an existing bookmark.
 * All fields are optional to support partial updates.
 */
export interface UpdateBookmarkInput {
  url?: string;
  title?: string;
  description?: string;
  tags?: string[];
}

/**
 * Paginated response for bookmark listing.
 */
export interface BookmarkPage {
  bookmarks: Bookmark[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

/**
 * Filter options for listing bookmarks.
 * Used as query parameters in API requests.
 */
export interface ListFilters {
  page?: number;
  pageSize?: number;
  search?: string;
  tag?: string;
}

/**
 * Error response structure from the API.
 */
export interface APIError {
  error: string;
  message?: string;
  statusCode?: number;
}

File: src/backend/internal/model/bookmark.go
package model

import (
	"time"
)

// Bookmark represents a saved web resource with metadata
type Bookmark struct {
	ID          string    `json:"id"`
	URL         string    `json:"url"`
	Title       string    `json:"title"`
	Description string    `json:"description,omitempty"`
	Tags        []string  `json:"tags,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// CreateBookmarkInput represents the input for creating a bookmark
type CreateBookmarkInput struct {
	URL         string   `json:"url"`
	Title       string   `json:"title"`
	Description string   `json:"description,omitempty"`
	Tags        []string `json:"tags,omitempty"`
}

// UpdateBookmarkInput represents the input for updating a bookmark
type UpdateBookmarkInput struct {
	URL         *string   `json:"url,omitempty"`
	Title       *string   `json:"title,omitempty"`
	Description *string   `json:"description,omitempty"`
	Tags        *[]string `json:"tags,omitempty"`
}

// ListFilters represents filter options for listing bookmarks
type ListFilters struct {
	Page     int
	PageSize int
	Tag      string
	Search   string
}

// BookmarkPage represents a paginated response of bookmarks
type BookmarkPage struct {
	Bookmarks []*Bookmark `json:"bookmarks"`
	Total     int64       `json:"total"`
	Page      int         `json:"page"`
	PageSize  int         `json:"page_size"`
	TotalPages int        `json:"total_pages"`
}

// BookmarkError represents an error response for bookmark operations
type BookmarkError struct {
	Error string `json:"error"`
	Code  int    `json:"code"`
}
